### ISSUE TITLE
Add support for moving resources from `gavinbunney/kubectl`

### ISSUE THREAD

**Issue by stevehipwell** [alekc/terraform-provider-kubectl#123]
With the release of Terraform v1.8 support for moving resources between providers has been added (this is also due to land in OpenTofu 1.8). It would be great if this provider could support moving resources from the original provider, `gavinbunney/kubectl`, to make switching this out much easier.

Looking at the code here it looks like you'd need to [migrate](https://developer.hashicorp.com/terraform/plugin/framework/migrating) to the [Terraform Plugin Framework](https://github.com/hashicorp/terraform-plugin-framework) to make use of the updated [state move functionality](https://developer.hashicorp.com/terraform/plugin/framework/resources/state-move), so that is probably the major constraint here.

**Comment by alekc** [comment]
Hmm, I am a bit confused here. Since this project has been originally forked from the gavinbunney's, all you need to do is to change the provider definition and run 
`terraform state replace-provider gavinbunney/kubectl alekc/kubectl`

?

**Comment by stevehipwell** [comment]
@alekc you can use the `terraform state replace-provider` command, but there are a number of issues with this pattern especially when working with modules.

- It requires all resources to be moved at one
  - This is particularly difficult to coordinate when there are multiple modules in a workspace
- It requires direct access to the state
  - In a least privileged environment this would require additional automation to be added
- It requires additional steps
  - Anyone taking a updated module would need to know to run the manual command
  - All modules and resources would need to move at once

We actually did the whole process of releasing all new module versions to align the providers, creating provider replace automation, and then documenting & supporting multiple teams to complete the process. It was painful.

**Comment by alekc** [comment]
Fair enough. I will have a look into it (can't promise timeframe though)

**Comment by alekc** [comment]
Bumping with usage stats from today (2026-05-20). This issue is still relevant. The gavinbunney provider remains actively pulled but is no longer being maintained, and the cohort still on it will increasingly be forced to migrate.

**Recent rate (best estimate from public registry / GitHub data)**:

| | gavinbunney/kubectl | alekc/kubectl |
| --- | --- | --- |
| Lifetime downloads | 186.4M | 55.4M |
| Public .tf files referencing it (GitHub code search) | 1,220 | 1,108 |
| Latest release | v1.19.0 (2025-01-10, 16 months ago) | v2.4.0 (2026-05-19) |
| Realistic recent rate (decay-adjusted estimate of the dominant version) | ~30-60K/day, ~210-420K/week, ~900K-1.8M/month | ~30-50K/day, ~210-350K/week, ~900K-1.5M/month |

Caveat: the Terraform Registry does not expose true weekly / monthly download counts publicly, so the recent-rate cells are upper-bounded estimates from lifetime-of-dominant-version divided by days-since-release. Code search hits and lifetime totals corroborate that current adoption is comparable.

**Forcing functions for the gavinbunney cohort**:

- v1.19.0 supports up to Kubernetes v1.32. Community PRs to add v1.33 ([gavinbunney/terraform-provider-kubectl#353](https://github.com/gavinbunney/terraform-provider-kubectl/pull/353)) and v1.34 ([gavinbunney/terraform-provider-kubectl#355](https://github.com/gavinbunney/terraform-provider-kubectl/pull/355)) were closed without merge. Anyone on k8s 1.33+ has to fork the repo themselves or switch providers.
- An open CVE tracker ([gavinbunney/terraform-provider-kubectl#356](https://github.com/gavinbunney/terraform-provider-kubectl/issues/356)) flags GO-2026-4337 / CVE-2025-68121 exposure as of March 2026 with no maintainer response since.

So the cohort that would benefit from native `moved` block support has both a k8s-version forcing function and a CVE forcing function. @stevehipwell's original objections to the `terraform state replace-provider` workaround (atomic-across-modules, requires state access, manual coordination) still apply.

**Comment by alekc** [comment]
Shipping in v3.0.0. The framework-side `kubectl_manifest` (PR #314) implements `MoveStateResource` for the `gavinbunney/kubectl` source, and PR #316 lands the workflow assertion + fingerprint recomputation needed for a clean cross-provider plan.

Recipe (Terraform 1.8+):

```hcl
terraform {
  required_providers {
    kubectl = { source = "alekc/kubectl" }
  }
}

moved {
  from = kubectl_manifest.my_app
  to   = kubectl_manifest.my_app_v3
}

resource "kubectl_manifest" "my_app_v3" {
  # same yaml_body / attributes
}
```

`terraform init -upgrade && terraform plan && terraform apply` migrates the state in place; after the apply commits the rename you can optionally rename back. README has the full step-by-step including the rationale for the transitional `_v3` address (Terraform's HCL validator rejects same-address `moved` blocks).

OpenTofu caveat: OpenTofu 1.11.x does not dispatch `MoveStateResource` on cross-provider transitions, so OpenTofu users should use `tofu state replace-provider gavinbunney/kubectl alekc/kubectl` instead. README documents both paths.

Closing on PR #316 merge.
